pkg/metrics/prometheus.go

```diff
@@ -16,60 +16,174 @@
+)
+
+type metrics string
+
+// prometheusCollector is a struct that holds the maps for histograms, gauges, and counters.
+type prometheusCollector struct {
+	// histograms is a map that holds the HistogramVec objects.
+	histograms map[string]*prometheus.HistogramVec
+	// gauges is a map that holds the GaugeVec objects.
+	gauges map[string]*prometheus.GaugeVec
+	// counters is a map that holds the CounterVec objects.
+	counters map[string]*prometheus.CounterVec
 }
 
-// NewPrometheus returns a new prometheus metrics implementation
-func NewPrometheus() Metrics {
-	return &prom{
-		counters: make(map[string]prometheus.Counter),
-		summary:  make(map[string]prometheus.Summary),
+// SetGauge sets a gauge metric with a given name, value, and labels.
+func (p *prometheusCollector) SetGauge(ctx context.Context, name string, value float64, labels ...string) error {
+	if _, ok := p.gauges[name]; !ok {
+		return fmt.Errorf("gauge '%s' not found", name)
 	}
+	p.gauges[name].WithLabelValues(labels...).Set(value)
+	return nil
 }
 
-// NewCounter creates a new counter metric
-func (p *prom) NewCounter(name string) error {
-	if _, ok := p.counters[name]; ok {
-		return fmt.Errorf("counter %s already exists", name)
+// AddHistogram increments a histogram metric with a given name, value, and labels.
+func (p *prometheusCollector) AddHistogram(ctx context.Context, name string, value float64, labels ...string) error {
+	if _, ok := p.histograms[name]; !ok {
+		return fmt.Errorf("histogram '%s' not found", name)
+	}
+	p.histograms[name].WithLabelValues(labels...).Observe(value)
+	return nil
+}
+
+// AddCounter increments a counter metric with a given name, value, and labels.
+// It returns an error if the counter is not found.
+func (p *prometheusCollector) AddCounter(ctx context.Context, name string, value float64, labels ...string) error {
+	if _, ok := p.counters[name]; !ok {
+		return fmt.Errorf("counter '%s' not found", name)
 	}
-	counter := prometheus.NewCounter(
-		prometheus.CounterOpts{
-			Name: fmt.Sprintf("%s_total", name),
-			Help: fmt.Sprintf("The total number of %s function calls", name),
-		})
-	prometheus.MustRegister(counter)
-	p.counters[name] = counter
+	p.counters[name].WithLabelValues(labels...).Add(value)
 	return nil
 }
 
-// IncrementCounter increments a counter metric
-func (p *prom) IncrementCounter(funcName string) {
-	p.counters[funcName].Inc()
+// MetricsHandler returns an http.Handler for the prometheus metrics.
+func (p *prometheusCollector) MetricsHandler() http.Handler {
+	return promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{
+		EnableOpenMetrics: true,
+	})
 }
 
-// NewSummary creates a new summary metric
-func (p *prom) NewSummary(name string) error {
-	if _, ok := p.summary[name]; ok {
-		return fmt.Errorf("summary %s already exists", name)
+// NewPrometheus creates a new prometheusCollector with empty sync.Maps for histograms, gauges, and counters.
+func NewPrometheus() MetricCollector {
```

Comment:
is it valuable here to take in a name of an instance so that metrics can be prefixed by their type (e.g. deps_dev_collector.xyz)

Reply:
I suggest that each package should have a prefix in its name.

 Changing the name of the `metrics` package is unnecessary.

Reply:
I agree. Included the option to take a name as a parameter. Thanks